### PR TITLE
Add support for reading binary files from stdin

### LINKED ISSUE #33: Add flag to accept raw image data in stdin
Example use case: `gpg --decrypt image.png.pgp | imv -r`

### PR DESCRIPTION
**Pull request by lutzee**
Fixes #33

### CODE REVIEW

**Comment by lutzee** [comment]
Updated the PR with a couple of minor fixes.

File: src/loader.c [hunk 1]
```diff
@@ -157,8 +159,25 @@ static void *imv_loader_bg_new_img(void *data)
   pthread_mutex_lock(&ldr->lock);
   char *path = strdup(ldr->path);
   pthread_mutex_unlock(&ldr->lock);
-
-  FREE_IMAGE_FORMAT fmt = FreeImage_GetFileType(path, 0);
+  FREE_IMAGE_FORMAT fmt = NULL;
+  FIBITMAP *bmp = NULL;
+  if(ldr->stdin_buffer != NULL && strcmp(path, "stdin") == 0){
```
Comment (eXeC64) [inline src/loader.c]: At this point, we should only do the `"stdin"` and `stdin_buffer` check and use that to set a variable, something like `int load_from_memory`.

Then each time we're touching `path` via FreeImage check `load_from_memory` and use that to decide between using `path`, and using `stdin_buffer.`

Doing it that way will simplify the logical changes made whilst letting animated gifs work with the stdin buffer.

**Comment by eXeC64** [comment]
There's a bunch of small + medium changes to be made I'm afraid. About to comment them in the diff.

File: src/loader.c [hunk 2]
```diff
@@ -36,9 +36,11 @@ static int is_thread_cancelled()
   return sigismember(&sigmask, SIGUSR1);
 }
 
-void imv_init_loader(struct imv_loader *ldr)
+void imv_init_loader(struct imv_loader *ldr, char* stdin_buffer, size_t stdin_buffer_size)
```
Comment (eXeC64) [inline src/loader.c]: `char *stdin_buffer` please. Asterisk sticks to the variable, not the type.

File: src/loader.c [hunk 3]
```diff
@@ -157,8 +159,30 @@ static void *imv_loader_bg_new_img(void *data)
...
+  FREE_IMAGE_FORMAT fmt = NULL;
+  FIBITMAP *bmp = NULL;
+
+  int load_from_memory = 0;
+  if(ldr->stdin_buffer != NULL && strcmp(path, "stdin") == 0){
+    load_from_memory = 1;
+  }
 
-  FREE_IMAGE_FORMAT fmt = FreeImage_GetFileType(path, 0);
+  if(load_from_memory == 1){
```
Comment (eXeC64) [inline src/loader.c]: At this point for both the path and the stdin case we should just be determining the filetype and setting fmt appropriately.

File: src/loader.c [hunk 4]
```diff
@@ -168,11 +192,16 @@ static void *imv_loader_bg_new_img(void *data)
 
   int num_frames = 1;
   FIMULTIBITMAP *mbmp = NULL;
-  FIBITMAP *bmp = NULL;
   int width, height;
   int raw_frame_time = 100; /* default to 100 */
 
   if(fmt == FIF_GIF) {
+    if(load_from_memory == 1){
```
Comment (eXeC64) [inline src/loader.c]: We can support gifs: `FreeImage_LoadMultiBitmapFromMemory()`. Just have to do the if memory/path check and use the appropriate function.

File: src/loader.c [hunk 5]
```diff
@@ -198,17 +227,20 @@ static void *imv_loader_bg_new_img(void *data)
       raw_frame_time = *(int*)FreeImage_GetTagValue(tag);
     }
     FreeImage_UnlockPage(mbmp, frame, 0);
-
   } else {
     /* Future TODO: If we load image line-by-line we could stop loading large
      * ones before wasting much more time/memory on them. */
-    FIBITMAP *image = FreeImage_Load(fmt, path, 0);
```
Comment (eXeC64) [inline src/loader.c]: This is where we call the `bla_Load` functions, depending on whether we're loading from memory/disk or not.

File: src/loader.c [hunk 6]
```diff
@@ -198,17 +227,20 @@ static void *imv_loader_bg_new_img(void *data)
...
-    free(path);
-    if(!image) {
-      imv_loader_error_occurred(ldr);
-      return 0;
+    FIBITMAP *image = NULL;
+    if(load_from_memory != 1){
```
Comment (eXeC64) [inline src/loader.c]: Also, space between `)` and `{` on ifs please, for consistency with the rest of the code.

File: src/main.c [hunk 7]
```diff
@@ -255,7 +274,11 @@ int main(int argc, char** argv)
 
   /* create our main classes on the stack*/
   struct imv_loader ldr;
-  imv_init_loader(&ldr);
+  if(g_options.binary_stdin){
```
Comment (eXeC64) [inline src/main.c]: If we pre-set `stdin_buffer` to NULL and `stdin_buffer_size` to 0 when we declare them, we don't need this if.

File: src/util.c [hunk 8]
```diff
@@ -19,6 +19,28 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 
 #include <fontconfig/fontconfig.h>
 
+char* load_img_stdin(size_t *lenout){
```
Comment (eXeC64) [inline src/util.c]: Opening brace on newline for functions please. Putting it on a newline helps separate the parameters from the body of the function.

**Comment by eXeC64** [comment]
Also, please update the manpage and usage information to document the new feature.

**Comment by eXeC64** [comment]
Solved by #76

**Pull request closed without merging**